05/index.html

```diff
@@ -0,0 +1,220 @@
+        }
+        100% {
+          transform: rotate(360deg);
+        }
+      }
+
+      #credits {
+        text-align: center;
+        font-size: 0.8rem;
+      }
+    </style>
+  </head>
+  <body>
+    <div id="container">
+      <div id="content">
+        <h1 style="text-align: center;">🔎 Emoji Search</h1>
+        <div id="search">
+          <input id="query" type="text" placeholder="Enter a query here"/>
+          <button id="search_button" disabled>Search</button>
+        </div>
+        <div id="emojis" class="output">
+          <div id="loader_box" hidden><span class="loader"></span></div>
+        </div>
+      </div>
+      <div id="credits">
+        🧶 Brought to you by <a href="https://serviceweaver.dev/">Service Weaver</a>.
+        💻 <a href="https://github.com/ServiceWeaver/workshops">Source Code</a>.
+      </div>
+    </div>
+
+    <script>
+      // strip removes all whitespace from the provided strings.
+      function strip(s) {
+        return s.replace(/\s+/g, '');
+      }
+
+      // emoji_span returns a freshly created span with the provided emoji as
+      // its contents. The returned span has the 'emoji' class and copies the
+      // emoji to the clipboard when clicked.
+      function emoji_span(emoji) {
+        const span = document.createElement('span');
+        span.innerText = emoji;
+        span.classList.add('emoji');
+        span.addEventListener('click', () => {
+          if (navigator.clipboard) {
+            navigator.clipboard.writeText(emoji);
+          }
+        });
+        return span;
+      }
+
+      // search queries the /{endpoint}?q={query} endpoint.
+      async function search(endpoint, query, aborter) {
+        const response = await fetch(`/${endpoint}?q=${query}`, {signal: aborter});
+        const text = await response.text();
+        if (response.ok) {
+          return text;
+        } else {
+          throw new Error(text);
+        }
+      }
+
+      function main() {
+        const query = document.getElementById('query');
+        const emojis = document.getElementById('emojis');
+        const button = document.getElementById('search_button');
+        const loader = document.getElementById('loader_box');
+        query.focus();
+
+        let controller; // for cancelling pending requests
+        let pending = 0; // number of pending requests
+        const displayed = new Set(); // emojis currently shown to the user
+        const perform_search = () => {
+          // Cancel any pending operations.
+          if (controller != undefined) {
+            controller.abort();
+          }
+          controller = new AbortController();
+
+          // Clear existing emojis. Be careful not to remove the loader.
+          while (emojis.children.length > 1) {
+            emojis.children[0].remove();
+          }
+          displayed.clear();
+
+          // Perform the search.
+          for (const endpoint of ['search', 'search_chatgpt']) {
```

Comment:
Yeah, since the amount of chatgpt code here is minimal and the user is not writing it, I figure having a single index.html file was more convenient than having them re-download a new file in part 10.